designs/images/cluster-controller-container-components.puml

```diff
@@ -0,0 +1,59 @@
+Container_Ext(workloadKubeApiServer, "Workload Cluster Kubernetes API server", "Go", "Provides read/write access to the kubernetes API objects stored in the etcd server")
+Container_Ext(ociRegistry, "OCI registry", "Public ECR or other", "OCI compliant registry to store container images and Helm charts")
+
+Container_Boundary(clusteController, "Cluster Controller") {
+    AddElementTag("port", $fontColor="#ffffbf", $borderColor="#ffffbf")
+    AddElementTag("busines logic", $fontColor="#01ffff", $borderColor="#01ffff")
+
+    Component(clusterReconciler, "Cluster Reconciler", "K8s controller-runtime Reconciler", "Validates input and orchestrates Cluster reconcile requests", $tags="busines logic")
+    Component(vsphereHandler, "vSphere Cluster Reconciler", "ProviderClusterReconciler", "Reconciles a vSphere Cluster to the desired state", $tags="busines logic")
+    Component(snowHandler, "Snow Cluster Reconciler", "ProviderClusterReconciler", "Reconciles a Snow Cluster to the desired state", $tags="busines logic")
+
+    Component(cilimHandler, "Cilium Reconciler", "", "Reonciles the Cilium CNI installation to the desired state", $tags="busines logic")
+
+    Component(client, "Kubernetes API client", "controller-runtime client.Client", "Kubernetes client provided by controller-runtime", $tags="port")
+    Component(remoteClient, "Remote Kubernetes API client", "controller-runtime client.Client", "Kubernetes client for external cluster authenticated with kubeconfig", $tags="port")
+
+    Component(helm, "Helm client", "Helm cli binary [Go]", "Client to interact with Helm charts", $tags="port")
+
+    Rel(vsphereHandler, cilimHandler, "Delegates")
+    Rel(snowHandler, cilimHandler, "Delegates")
+
+    Rel(clusterReconciler, vsphereHandler, "Delegates")
+    Rel(clusterReconciler, snowHandler, "Delegates")
+
+    Rel(vsphereHandler, client, "Reads/writes CAPI objects through")
+    Rel(snowHandler, client, "Reads/writes CAPI objects through")
+    Rel(clusterReconciler, client, "Reads objects through")
+    Rel(cilimHandler, remoteClient, "Reads/writes Cilium DS/deployment and CRDs")
+
+    Rel(cilimHandler, helm, "Generates Cilium resources in yaml")
+}
+
+Container(capiControllers, "CAPI controllers", "Go", "Reoncile workload clusters")
+Rel(snowHandler, capiControllers, "Creates/updates snow clusters", $tags="indirect relantionship")
+Rel(vsphereHandler, capiControllers, "Creates/updates vSphere clusters", $tags="indirect relantionship")
+Rel(capiControllers, cilimHandler, "Marks control plane as ready", $tags="indirect relantionship")
+Rel(capiControllers, workloadKubeApiServer, "Creates", $tags="indirect relantionship")
+
+Rel(user, kubeApiServer, "Writes EKS-A API objects", "JSON/HTTPS")
+
+Rel(clusteController, kubeApiServer, "Subscribes to API events", "JSON/HTTPS")
```

Comment:
On component diagrams there's no need to depict higher level relationships like container to container. The focus should be on _what component_ communicates with an external container/system.

Reply:
Yeah, that's fair.

I felt similarly, however, I wanted to represent this relationship, I think it's useful to understand the input of events to the components inside the container.

I saw 3 options:
* What I have here
* Start the relationship in the cluster reconciler. Which is directionally correct but technically incorrect, since the subscription is managed by the controller itself (you define the subscription when registering the controller with the controller manager).
* Add another component, like a "generic" `controller-runtime`. This would perform the subscription and send the reconcile request to the cluster reconciler. This seemed going a bit too deep into how the controller-runtime framework works. 

What do you think?

Reply:
I think the relationship belongs on a container diagram, not a component diagram :).

As is, this denotes a component in the container talks to kube API but doesn't tell us what component and that defeats the point of diving a level deeper than the container.